src/server.py

```diff
@@ -0,0 +1,188 @@
+    """Download Jenkins agent JAR executable from server.
+
+    Args:
+        server_url: The Jenkins server URL address.
+        connectable_container: The connectable agent container.
+
+    Raises:
+        AgentJarDownloadError: If an error occurred downloading the JAR executable.
+    """
+    try:
+        res = requests.get(f"{server_url}/jnlpJars/agent.jar", timeout=300)
+        res.raise_for_status()
+    except (requests.HTTPError, requests.Timeout, requests.ConnectionError) as exc:
+        logger.error("Failed to download agent JAR executable from server, %s", exc)
+        raise AgentJarDownloadError(
+            "Failed to download agent JAR executable from server."
+        ) from exc
+
+    connectable_container.push(
+        path=AGENT_JAR_PATH, make_dirs=True, source=res.content, user=USER, group=GROUP
+    )
+
+
+def validate_credentials(
+    agent_name: str, credentials: Credentials, connectable_container: ops.Container
+) -> bool:
+    """Check if the credentials can be used to register to the server.
+
+    Args:
+        agent_name: The Jenkins agent name.
+        credentials: Server credentials required to register to Jenkins server.
+        connectable_container: The Jenkins agent workload container.
+
+    Returns:
+        True if credentials and agent_name pairs are valid, False otherwise.
+    """
+    # IMPORTANT: add random jitter to prevent parallel execution.
+    # It's okay to use random since it's not used for sensitive data.
+    time.sleep(random.random())  # nosec
+    proc: ops.pebble.ExecProcess = connectable_container.exec(
+        [
+            "java",
+            "-jar",
+            str(AGENT_JAR_PATH),
+            "-jnlpUrl",
+            f"{credentials.address}/computer/{agent_name}/slave-agent.jnlp",
+            "-workDir",
+            str(JENKINS_WORKDIR),
+            "-noReconnect",
+            "-secret",
+            credentials.secret,
+        ],
+        timeout=5,
+        user=USER,
+        group=GROUP,
+        working_dir=str(JENKINS_WORKDIR),
+        combine_stderr=True,
+    )
+    # The process will exit due to connection failure(invalid credentials) or timeout.
+    # Check for successful connection log from the stdout.
+    connected = False
+    terminated = False
+    # The proc.stdout is iterable according to process.exec documentation
+    for line in proc.stdout:  # type: ignore
+        if "INFO: Connected" in line:
+            connected = True
+        if "INFO: Terminated" in line:
+            terminated = True
+
+    return connected and not terminated
+
+
+def is_registered(connectable_container: ops.Container) -> bool:
+    """Check if the given agent instance is already up and running.
+
+    Args:
+        connectable_container: The Jenkins agent workload container.
+
+    Returns:
+        Whether the Jenkins agent is already registered.
+    """
+    return connectable_container.exists(str(AGENT_READY_PATH))
```

Comment:
My intention with this function was to abstract/wrap the business logic as a function(since originally I wanted to use the `netstat` method). The alternative could be to inline it and have a comment explaining it, which sounds more suitable now. Updating, thank you for the comment!